server/src/main/java/org/elasticsearch/cluster/node/DiscoveryNodes.java

```diff
@@ -350,29 +353,31 @@ public DiscoveryNode resolveNode(String node) {
                     if (index != -1) {
                         String matchAttrName = nodeId.substring(0, index);
                         String matchAttrValue = nodeId.substring(index + 1);
-                        if (DiscoveryNodeRole.DATA_ROLE.roleName().equals(matchAttrName)) {
-                            if (Booleans.parseBoolean(matchAttrValue, true)) {
-                                resolvedNodesIds.addAll(dataNodes.keys());
-                            } else {
-                                resolvedNodesIds.removeAll(dataNodes.keys());
-                            }
-                        } else if (DiscoveryNodeRole.MASTER_ROLE.roleName().equals(matchAttrName)) {
-                            if (Booleans.parseBoolean(matchAttrValue, true)) {
-                                resolvedNodesIds.addAll(masterNodes.keys());
+                        if (DiscoveryNodeRole.BUILT_IN_ROLES.stream()
+                            .map(DiscoveryNodeRole::roleName)
+                            .anyMatch(s -> s.equals(matchAttrName))) {
+                            final DiscoveryNodeRole role = DiscoveryNode.getRoleFromRoleName(matchAttrName);
+                            final Predicate<Set<DiscoveryNodeRole>> predicate;
+                            if (role.equals(DiscoveryNodeRole.DATA_ROLE)) {
+                                // if the node has *any* role that can contain data, then it matches the data attribute
+                                predicate = s -> s.stream().anyMatch(DiscoveryNodeRole::canContainData);
+                            } else if (role.canContainData()) {
+                                // if the node has the matching data_ role, or the generic data role, then it matches the data_ attribute
+                                predicate = s -> s.stream().anyMatch(r -> r.equals(role) || r.equals(DiscoveryNodeRole.DATA_ROLE));
                             } else {
-                                resolvedNodesIds.removeAll(masterNodes.keys());
+                                // the role is not a data role, we require an exact match (e.g., ingest)
+                                predicate = s -> s.contains(role);
                             }
-                        } else if (DiscoveryNodeRole.INGEST_ROLE.roleName().equals(matchAttrName)) {
+                            final Function<String, Boolean> mutation;
                             if (Booleans.parseBoolean(matchAttrValue, true)) {
-                                resolvedNodesIds.addAll(ingestNodes.keys());
+                                mutation = resolvedNodesIds::add;
                             } else {
-                                resolvedNodesIds.removeAll(ingestNodes.keys());
+                                mutation = resolvedNodesIds::remove;
                             }
-                        } else if (DiscoveryNode.COORDINATING_ONLY.equals(matchAttrName)) {
```

Comment:
I believe this case is no longer handled. `COORDINATING_ONLY` is not included in `BUILTIN_ROLES`, but is allowed in node specifications.

`DiscoveryNodesTests.testCoordinatorOnlyNodes` fails like 1/3 of the time.